### ISSUE TITLE
Phase 8.6.6: ML-Based Ranking - Gradient Boosted Trees Model

### ISSUE THREAD

**Issue by jpequegn** [jpequegn/ai-coach#183]
## Overview
Implement ML-based recommendation ranking using gradient boosted trees to predict recommendation effectiveness and completion probability.

**Parent Issue**: #134 - Phase 8.6: Smart Recommendation Features & ML Enhancements
**Priority**: Phase 4 (ML/Advanced) - Requires sufficient training data
**Estimated Time**: 3-4 days

## Objectives

Replace or augment rule-based scoring with machine learning models that learn from historical user interactions to predict which recommendations will be most effective.

## Prerequisites

- Minimum 1000 completed recommendations with outcomes across 100+ users
- At least 30 days of historical data
- Recommendation effectiveness data from issue #133

### 1. Feature Engineering

- [ ] Create RecommendationFeatures struct:
  ```rust
  pub struct RecommendationFeatures {
      // User features
      pub user_completion_rate: f64,
      pub user_avg_rating: f64,
      pub user_activity_level: f64,
      pub user_weeks_active: i32,
      pub user_category_completion_rate: HashMap<String, f64>,

      // Recommendation features
      pub template_effectiveness: f64,
      pub template_times_suggested: i32,
      pub template_times_completed: i32,
      pub template_avg_rating: f64,
      pub category_preference: f64,
      pub time_since_last_similar: i32,
      pub difficulty_match: f64,

      // Context features
      pub recovery_score: f64,
      pub hrv_deviation: f64,
      pub sleep_quality: f64,
      pub stress_level: f64,
      pub time_of_day: f64, // 0-23 encoded
      pub day_of_week: f64, // 0-6 encoded
      pub season: String,

      // Interaction features
      pub user_template_previous_rating: Option<f64>,
      pub similar_users_effectiveness: f64,
  }
  ```

- [ ] Feature extraction service:
  - extract_user_features(user_id)
  - extract_recommendation_features(template_id)
  - extract_context_features(recovery_score, context)
  - extract_interaction_features(user_id, template_id)

- [ ] Feature normalization and scaling:
  - StandardScaler for continuous features
  - One-hot encoding for categorical features
  - Handle missing values appropriately

### 2. Training Data Preparation

- [ ] Create training dataset from historical data:
  ```sql
  SELECT
      ur.user_id,
      ur.recommendation_template_id,
      ur.status,
      ro.effectiveness_score as target,
      ur.effectiveness_rating,
      -- Join all feature tables
  FROM user_recommendations ur
  LEFT JOIN recommendation_outcomes ro ON ur.id = ro.user_recommendation_id
  WHERE ur.status IN ('completed', 'skipped')
  AND ur.shown_at > NOW() - INTERVAL '90 days';
  ```

- [ ] Define prediction targets:
  - Primary: Combined effectiveness score (0-1)
  - Secondary: Completion probability (0-1)
  - Tertiary: User rating prediction (1-5)

- [ ] Train/test split:
  - 80% training, 20% testing
  - Time-based split (train on older data, test on recent)
  - Stratify by user to ensure representation

### 3. Model Training

- [ ] Implement GBT model using linfa or smartcore:
  ```rust
  use linfa::prelude::*;
  use linfa_trees::DecisionTree;

  pub struct MLRankingModel {
      model: GradientBoostedTrees,
      feature_scaler: StandardScaler,
      feature_names: Vec<String>,
      model_version: String,
      trained_at: DateTime<Utc>,
  }
  ```

- [ ] Model hyperparameters:
  - Number of trees: 100-300
  - Max depth: 5-7
  - Learning rate: 0.01-0.1
  - Min samples split: 10-20
  - Regularization: L1/L2

- [ ] Training process:
  1. Load and prepare training data
  2. Extract features for all samples
  3. Train GBT model with cross-validation
  4. Evaluate on test set
  5. Save model artifacts

- [ ] Model evaluation:
  - RMSE for effectiveness prediction
  - Accuracy for completion prediction
  - ROC-AUC for ranking quality
  - Feature importance analysis

### 4. Model Serving

- [ ] Create MLModelService:
  ```rust
  impl MLModelService {
      pub async fn load_model(version: &str) -> Result<MLRankingModel>;

      pub async fn predict_effectiveness(
          &self,
          features: &RecommendationFeatures
      ) -> Result<f64>;

      pub async fn rank_recommendations(
          &self,
          user_id: Uuid,
          templates: Vec<RecommendationTemplate>,
          context: RecommendationContext
      ) -> Result<Vec<ScoredRecommendation>>;
  }
  ```

- [ ] Model loading and caching:
  - Load model on service startup
  - Cache in memory for fast inference
  - Hot-reload on model updates

- [ ] Batch prediction optimization:
  - Vectorize feature extraction
  - Batch predict for multiple recommendations
  - Target <50ms latency for ranking

### 5. Model Versioning and Deployment

- [ ] Model storage:
  ```
  models/
    recommendation_ranker/
      v1.0.0/
        model.bin
        scaler.bin
        metadata.json
      v1.1.0/
        model.bin
        scaler.bin
        metadata.json
  ```

- [ ] Metadata tracking:
  ```json
  {
    "version": "1.0.0",
    "trained_at": "2025-01-15T10:30:00Z",
    "training_samples": 5000,
    "test_rmse": 0.08,
    "test_accuracy": 0.73,
    "features": ["user_completion_rate", ...],
    "hyperparameters": {...}
  }
  ```

- [ ] Deployment strategy:
  - Blue-green deployment for model updates
  - A/B testing capability
  - Rollback mechanism

### 6. Integration with Recommendation Engine

- [ ] Update RecommendationEngine:
  - Add ML scoring option alongside rule-based
  - Ensemble approach: blend ML and rule-based scores
  - Feature flag for ML scoring activation

- [ ] Scoring strategies:
  - Pure ML: 100% ML model predictions
  - Hybrid: 70% ML + 30% rule-based
  - Fallback: Rule-based if ML unavailable

### 7. Continuous Learning Pipeline

- [ ] Automated retraining:
  - Daily data collection
  - Weekly model retraining
  - Automatic deployment if improvement >5%

- [ ] Performance monitoring:
  - Track prediction accuracy in production
  - Monitor feature drift
  - Alert on model degradation

- [ ] Data pipeline:
  ```rust
  pub struct TrainingPipeline {
      pub collect_training_data() -> Result<Dataset>;
      pub train_model(data: Dataset) -> Result<MLRankingModel>;
      pub evaluate_model(model: &MLRankingModel) -> Result<ModelMetrics>;
      pub deploy_model(model: MLRankingModel) -> Result<()>;
  }
  ```

### 8. API Endpoints

- [ ] POST /api/v1/recovery/recommendations/ml-ranked
  - Body: Same as current endpoint
  - Returns: ML-ranked recommendations with prediction scores

- [ ] GET /api/v1/ml/models/ranking
  - Returns: Current model version and metrics

- [ ] POST /api/v1/ml/models/ranking/retrain (admin only)
  - Triggers model retraining
  - Returns: Training job status

### 9. Feature Importance and Explainability

- [ ] Calculate and expose feature importance:
  - SHAP values for individual predictions
  - Global feature importance ranking
  - User-specific feature contributions

- [ ] Explainability API:
  ```json
  {
    "recommendation_id": "uuid",
    "predicted_effectiveness": 0.87,
    "top_factors": [
      {"feature": "user_completion_rate", "contribution": 0.3},
      {"feature": "recovery_score", "contribution": 0.25},
      {"feature": "template_effectiveness", "contribution": 0.2}
    ]
  }
  ```

### 10. Training Scripts

- [ ] Create training binary:
  ```bash
  cargo run --bin train_recommendation_model -- \
    --data-path data/training.csv \
    --output-path models/recommendation_ranker/v1.0.0/ \
    --n-trees 200 \
    --max-depth 6 \
    --learning-rate 0.05
  ```

- [ ] Evaluation script:
  ```bash
  cargo run --bin evaluate_recommendation_model -- \
    --model-path models/recommendation_ranker/v1.0.0/ \
    --test-data data/test.csv
  ```

## Acceptance Criteria

- [ ] ML model trained with >1000 samples
- [ ] Model achieves >60% accuracy on completion prediction
- [ ] Model RMSE <0.15 for effectiveness prediction
- [ ] ML ranking integrated with recommendation engine
- [ ] Model versioning and deployment working
- [ ] Feature importance tracked and exposed
- [ ] Continuous retraining pipeline functional
- [ ] API endpoints return ML-ranked recommendations
- [ ] A/B testing capability implemented

## Testing Requirements

- [ ] Unit tests for feature extraction
- [ ] Test model training pipeline
- [ ] Test model loading and inference
- [ ] Test with synthetic data
- [ ] Validate on historical data
- [ ] Integration tests for API endpoints
- [ ] Performance tests for ranking latency

## Success Metrics

- Model accuracy: >60% on completion prediction
- Effectiveness RMSE: <0.15
- Ranking quality: >70% of top-3 recommendations completed
- User satisfaction: ML recommendations rated ≥ rule-based
- Inference latency: <50ms for batch ranking
- Model improvement: +10% effectiveness over rule-based

## Dependencies

- Requires: #129, #130, #131, #132, #133 (all completed)
- Requires: Minimum 1000 completed recommendations with outcomes
- Parent: #134

## Library Recommendations

- ML Framework: `linfa` (native Rust) or `smartcore`
- Gradient Boosting: `linfa-trees` or external LightGBM bindings
- Feature scaling: `linfa-preprocessing`
- Model serialization: `bincode` or `serde`
- SHAP values: External Python service or port to Rust

## Notes

- Start with simple models before complex architectures
- Collect sufficient training data before implementation (3-6 months)
- Monitor for overfitting on small datasets
- Consider cold start problem for new users
- Plan for computational resources for training
- Implement feature store for consistent feature generation
- Consider online learning for real-time model updates
